File: src/main/java/cn/don9cn/blog/autoconfigs/shiro/core/MyPermission.java
package cn.don9cn.blog.autoconfigs.shiro.core;

import org.apache.shiro.authz.Permission;

/**
 * @Author: liuxindong
 * @Description: 自定义shiro Permission的实现
 * @Create: 2017/10/17 15:27
 * @Modify:
 */
public class MyPermission implements Permission{

    private final String requestUrl;

    private final String requestMethod;

    public MyPermission(String requestUrl, String requestMethod) {
        this.requestUrl = requestUrl;
        this.requestMethod = requestMethod;
    }

    /**
     * 权限验证,返回值为true时说明当前的请求与用户的权限相匹配
     * @param permission 用户请求的资源路径
     * @return
     */
    @Override
    public boolean implies(Permission permission) {
        MyPermission request = (MyPermission) permission;
        if(this.requestMethod.equals("ALL")){
            if(request.getRequestUrl().startsWith(this.getRequestUrl()))
                return true;
        }else{
            if(request.getRequestUrl().equals(this.requestUrl) && request.getRequestMethod().equals(this.requestMethod))
                return true;
        }
        return false;
    }

    public String getRequestUrl() {
        return requestUrl;
    }

    public String getRequestMethod() {
        return requestMethod;
    }
}

File: src/main/java/cn/don9cn/blog/autoconfigure/shiro/core/MyPermission.kt
package cn.don9cn.blog.autoconfigure.shiro.core

import org.apache.shiro.authz.Permission
import java.io.Serializable

/**
 * 自定义shiro Permission的实现
 */
class MyPermission(private val requestUrl: String, private val requestMethod: String) : Permission, Serializable {

    companion object {
        private const val serialVersionUID = 1L
    }

    /**
     * 权限验证,返回值为true时说明当前的请求与用户的权限相匹配
     * @param permission 用户请求的资源路径
     * @return
     */
    override fun implies(permission: Permission): Boolean {
        val request = permission as MyPermission
        if (this.requestMethod == "ALL") {
            if (request.requestUrl.startsWith(this.requestUrl))
                return true
        } else {
            if (request.requestUrl == this.requestUrl && request.requestMethod == this.requestMethod)
                return true
        }
        return false
    }

}